src/utils/downloadAsDocx/downloadAsDocx.ts

```diff
@@ -23,46 +19,42 @@ const downloadConfiguration = {
   pageNumber: true,
 }
 
-const createLectureBlob = async (lecture: Data<LectureTwoLevelsDeep>) => {
-  const sourceHTML = ReactDOMServer.renderToString(
-    LectureDocxDownload({ lecture })
-  )
-  const newHtml = await processHTMLString(sourceHTML, lecture.attributes.Title)
-  const lectureBlob = await HTMLtoDOCX(
-    newHtml,
-    undefined,
-    downloadConfiguration,
-    footer
-  )
-  return lectureBlob
-}
-
-export const handleCourseDocxDownload = async (
-  course: Data<CourseThreeLevelsDeep>
-) => {
-  const zip = new JSZip()
-  for (const lecture of course.attributes.Lectures.data) {
-    try {
-      const lectureBlob = await createLectureBlob(lecture)
-      zip.file(`${lecture.attributes.Title}.docx`, lectureBlob)
-    } catch (error) {
-      console.error(error)
-      zip.file(
-        `FAILED_${lecture.attributes.Title}.docx`,
-        `Docx generation failed for lecture ${lecture.attributes.Title}`
-      )
+const createBlob = async (data: Data<DownloadableContent>) => {
+  if (isCourseThreeLevelsDeep(data)) {
+    const zip = new JSZip()
+    for (const lecture of data.attributes.Lectures.data) {
+      try {
+        const { blob } = await createBlob(lecture)
+        zip.file(`${lecture.attributes.Title}.docx`, blob)
+      } catch (error) {
+        console.error(error)
+        zip.file(
+          `FAILED_${lecture.attributes.Title}.docx`,
+          `Docx generation failed for lecture ${lecture.attributes.Title}`
+        )
+      }
     }
+    const blob = await zip.generateAsync({ type: 'blob' })
+    return { blob, title: `${data.attributes.Title}.zip` }
+  } else {
+    const sourceHTML = ReactDOMServer.renderToString(DocxDownload({ data }))
+    const newHtml = await processHTMLString(sourceHTML, data.attributes.Title)
+    const blob = await HTMLtoDOCX(
+      newHtml,
+      undefined,
+      downloadConfiguration,
+      footer
+    )
+    return { blob, title: `${data.attributes.Title}.docx` }
```

Comment:
Cool refactor! I'm not sure I think it has better readability, but it sure is more clever and clean. Especially given the fact that this function is run from multiple places.

Reply:
No the readability is still tricky, but yes I thought removing some code duplication could help if we decide to change the docx file handling further down the line.